packages/widgets/src/blocks/legacy-widget/edit/preview.js

```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { useRefEffect } from '@wordpress/compose';
-import { addQueryArgs } from '@wordpress/url';
-import { useState } from '@wordpress/element';
+import { useState, useEffect } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
+	const [ srcDoc, setSrcDoc ] = useState( '' );
+	const [ isPreviewFetched, setIsPreviewFetched ] = useState( false );
+
+	const abortController =
+		typeof window.AbortController === 'undefined'
+			? undefined
+			: new window.AbortController();
+
+	async function fetchPreviewHTML() {
+		const restRoute = `/wp/v2/widget-types/${ idBase }/render`;
+		return await apiFetch( {
+			path: restRoute,
+			method: 'POST',
+			signal: abortController?.signal,
+			data: {
+				id_base: idBase,
+				instance,
+			},
+		} );
+	}
+
+	useEffect( () => {
+		if ( ! isPreviewFetched ) {
+			setIsPreviewFetched( true );
+			fetchPreviewHTML()
+				.then( ( response ) => {
+					setSrcDoc( response.preview );
+				} )
+				.catch( ( error ) => {
+					if ( error.name === 'AbortError' ) {
+						// We don't want to log abort errors.
+						return;
+					}
+					window.console.error(
+						`An error occurred while trying to fetch preview: ${ error.message }`,
+						error
+					);
+				} );
+		}
+
+		return () => {
+			if ( typeof abortController !== 'undefined' ) {
+				abortController.abort();
+			}
+		};
+	}, [] );
```

Comment:
We'll want to re-run this effect whenever `idBase` or `instance` changes.

```suggestion
	}, [ idBase, instance ] );
```

Reply:
Good catch.
Fixed in 982c2053d656220b0047641133763d211b6d6868